src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -70,103 +94,201 @@ class MacData {
+                    params.getSalt(), params.getIterationCount(), len));
         } else {
-            this.iterations = 1;
+            hmac = macAlgorithm;
+            m = Mac.getInstance(hmac);
+            keySpec = new PBEKeySpec(password);
+            skf = SecretKeyFactory.getInstance("PBE");
+            pbeKey = skf.generateSecret(keySpec);
+        }
+        keySpec.clearPassword();
+
+        try {
+            if (macAlgorithm.startsWith("pbewith")) {
+                m.init(pbeKey);
+            } else {
+                m.init(pbeKey, params);
+            }
+        } finally {
+            sun.security.util.KeyUtil.destroySecretKeys(pbeKey);
+        }
+        m.update(data);
+        return m;
+    }
+
+    void processMacData(char[] password, byte[] data)
+            throws InvalidKeySpecException,
+            NoSuchAlgorithmException, UnrecoverableKeyException,
+            InvalidKeyException, InvalidAlgorithmParameterException {
+        Mac m;
+        byte[] macResult;
+
+        m = getMac(this.macAlgorithm, password,
+                new PBEParameterSpec(this.macSalt, this.iterations),
+                data, this.kdfHmac, this.hmac, this.keyLength);
+        macResult = m.doFinal();
+
+        if (debug != null) {
+            debug.println("Checking keystore integrity " +
+                    "(" + m.getAlgorithm() + " iterations: "
+                    + this.iterations + ")");
+        }
+
+        if (!MessageDigest.isEqual(this.digest, macResult)) {
+            throw new UnrecoverableKeyException("Failed PKCS12" +
+                    " integrity checking");
         }
     }
 
-    MacData(String algName, byte[] digest, byte[] salt, int iterations)
-        throws NoSuchAlgorithmException
-    {
-        if (algName == null)
-           throw new NullPointerException("the algName parameter " +
-                                               "must be non-null");
-
-        AlgorithmId algid = AlgorithmId.get(algName);
-        this.digestAlgorithmName = algid.getName();
-        this.digestAlgorithmParams = algid.getParameters();
-
-        if (digest == null) {
-            throw new NullPointerException("the digest " +
-                                           "parameter must be non-null");
-        } else if (digest.length == 0) {
-            throw new IllegalArgumentException("the digest " +
-                                                "parameter must not be empty");
+    /*
+     * Calculate MAC using HMAC algorithm (required for password integrity)
+     *
+     * Hash-based MAC algorithm combines secret key with message digest to
+     * create a message authentication code (MAC)
+     */
+    static byte[] calculateMac(char[] passwd, byte[] data,
+            String macAlgorithm, int macIterationCount, byte[] salt)
+            throws IOException, NoSuchAlgorithmException {
+        final byte[] mData;
+        final PBEParameterSpec params;
+        String algName = "PBMAC1";
```

Comment:
No need to assign a value now. Assign it in the `if (macAlgorithm.startsWith("pbewith"))` block.